### DISCUSSION TITLE
Response is not a reader(stream)

### DISCUSSION THREAD

**Discussion by axetroy** [vlang/v#12675] [Ideas]
Response in `net.http` is not a reader

so, many stream-based functions cannot work.

**Comment by JalonSolov** [comment]
Well, no, Response is a struct:
```v
pub struct Response {
pub mut:
	text         string
	header       Header
	status_code  int
	status_msg   string
	http_version string
}
```
It currently reads the entire body into the `text` field.  What you may be asking for is to make `text` a Reader instead of `string`.

**Comment by axetroy** [comment]
> Well, no, Response is a struct:
> 
> ```v
> pub struct Response {
> pub mut:
> 	text         string
> 	header       Header
> 	status_code  int
> 	status_msg   string
> 	http_version string
> }
> ```
> 
> It currently reads the entire body into the `text` field. What you may be asking for is to make `text` a Reader instead of `string`.

I found the source code of HTTP in here.

https://github.com/vlang/v/blob/82e6d6e51d81602c80a220f7b04192c69511c010/vlib/net/http/request.v#L150-L168

When the request is over, read the text returned by the request from here, and then generate a Response

This step should be reserved for the user, otherwise, all the data will be loaded into the memory, and functions like the download progress bar cannot be done with this

**Comment by JalonSolov** [comment]
I'd rather have both.  If you're doing something where you know the response data won't be "big" (which is relative, I know), there's no need of the overhead of extra processing to stream it, just let the routine hand it back.

However, sometimes it is good to be able to stream the data, especially if it is really "big", or you only want to save a few bits out of the middle - instead of loading the whole thing first, copying pieces out, then throwing away what was downloaded.

**Comment by axetroy** [comment]
It V got a `Reader/Writer` interface in std?

Maybe I can make a PR

**Comment by santerijps** [comment]
Has this been taken into development yet? I would really like to see this feature in V.

**Comment by ttytm** [comment]
Would be great to have this! And a good base for more advanced features.
E.g. files 100mb+ throw Mem aloc. warnings. 
```v
import net.http
import time

fn main() {
	start := time.ticks()

	// resp := http.get('https://speed.hetzner.de/100MB.bin')!
	http.download_file('https://speed.hetzner.de/100MB.bin', '100MB.bin')!

	println('Took: ${f64(time.ticks() - start) / 1000:.2f}s')
}
```
```
GC Warning: Repeated allocation of very large block (appr. size 104861696):
        May lead to memory leak and poor performance
```

In go the response body is a reader which makes it a breeze to read a chunk at a time.
